### PR TITLE
PBIS AD login integration

### PR DESCRIPTION
**Pull request by tekeri**
###### Motivation for this change
Bring Active Directory authentication to NixOS.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/security/pbis.nix [hunk 1]
```diff
@@ -0,0 +1,148 @@
+{ config, lib, pkgs, ... }:
```
Comment (aanderse) [inline nixos/modules/services/security/pbis.nix]: I would encourage you to add yourself as a maintainer to this module, if you are interested in doing so.

**Review by aanderse — CHANGES_REQUESTED** [review]
I've taken a first pass at review of your module. A number of my comments are just minor points, convention, etc... but there are a few material points I look forward to getting your feedback on. I will test this on one of my servers at some point.

Out of curiosity what are the benefits you saw in this solution compared to others available?

Thanks for contributing this module!

File: nixos/modules/services/security/pbis.nix [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+      '';
+
+      postStart = with cfg.domainJoin; ''
+        ${pbis}/bin/pbis-status | grep "Unknown" >/dev/null && {
+          ${pbis}/bin/domainjoin-cli join ${toString extraOptions} \
+            ${escapeShellArg domain} \
+            ${escapeShellArg userName} \
+            ${optionalString (passwordPath != null) "$(cat ${passwordPath})"} \
+            >/dev/null 2>&1 || true
+            # ^ignore errors: pbis's stateful behavior (change PAM module and conf) cause some errors
```
[the remark this answers is not in the document]
Comment (tekeri) [reply nixos/modules/services/security/pbis.nix]: Sadly, no.

File: nixos/modules/services/security/pbis.nix [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+      # taken from sssd
+      before = [ "systemd-user-sessions.service" "nss-user-lookup.target" ];
+      after = [ "network-online.target" "nscd.service" ];
+      requires = [ "network-online.target" "nscd.service" ];
+      wants = [ "nss-user-lookup.target" ];
+      restartTriggers = [ config.environment.etc."nscd.conf".source ];
+
+      script = ''
+        set -x
+        # This is evil but I could not find any workaround that
```
[the remark this answers is not in the document]
Comment (tekeri) [reply nixos/modules/services/security/pbis.nix]: Less likely, as they are busy and cannot update their bundled dependencies even.
Instead, made `/opt/pbis` private to the service.

File: nixos/modules/services/security/pbis.nix [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+      example = { RequireMembershipOf = ''MyOrg\MyTeam''; AssumeDefaultDomain = "true"; };
+      description = ''
+        Sequence of setup subcommands to feed to `$${cfg.package}/bin/config`
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+
+    systemd.services.lwsmd = {
```
[the remark this answers is not in the document]
Comment (tekeri) [reply nixos/modules/services/security/pbis.nix]: Tried `DynamicUser` and found that
`chown("/var/lib/pbis/db", 0, 0) = -1 EPERM (Operation not permitted)`
... and it also create new user's home directory.

**Comment by tekeri** [comment]
> Out of curiosity what are the benefits you saw in this solution compared to others available?

Compatibility. PBIS derives uid/gid from users' SID.
Pros: no need to manage IDs.
Cons: it feels like vendor lock-in when I consider uid/gid consistency across different systems or host-container.

File: nixos/modules/services/security/pbis.nix [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+          $REGSHELL import "$i"
+          $REGSHELL cleanup "$i"
+        done
+        $REGSHELL set_value '[HKEY_THIS_MACHINE\Services\lwio\Parameters\Drivers]' 'Load' 'rdr'
+        $REGSHELL set_value '[HKEY_THIS_MACHINE\Services\lsass]' 'Dependencies' 'netlogon lwio lwreg rdr'
+        $REGSHELL set_value '[HKEY_THIS_MACHINE\Services\eventlog]' 'Dependencies' ""
+        $REGSHELL set_value '[HKEY_THIS_MACHINE\Services\dcerpc]' 'Arguments' ""
+        $REGSHELL set_value '[HKEY_THIS_MACHINE\Services\lsass\Parameters\Providers\ActiveDirectory]' Path '${cfg.package}/lib/lsa-provider/ad_open.so'
+        ${pbis}/bin/lwsm shutdown
+        exec ${pbis}/bin/lwsmd --start-as-daemon
```
[the remark this answers is not in the document]
Comment (tekeri) [reply nixos/modules/services/security/pbis.nix]: It did not work.

**Comment by aanderse** [comment]
> Pros: no need to manage IDs.

Oh... you could add `map passwd uidNumber objectSid:S-1-5-21-3268520485-1580165246-4160084142` to `users.ldap.daemon.extraConfig` and have the same result. Is that the only reason?

**Comment by tekeri** [comment]
It is not realistic to manage uids by hand every time people join or leave...
And pbis is needed to make NixOS as Ubuntu replacement.

**Comment by aanderse** [comment]
@tekeri I think you misunderstand... what that configuration does is compute a **unique** `uid` for each and every user based on the users underlying `objectSid` from AD. That way you don't have to assign or manage a `uid` for anyone as all `uid` values are a calculated function, unique across all members, and consistent across all machines.

See https://www.markturner.net/2019/09/27/ad-ldap-authentication-on-linux-hosts/ for a good explanation.

**Comment by tekeri** [comment]
@aanderse Umm.. I'm not sure that configuration utilize same hash function defined in pbis? https://github.com/BeyondTrust/pbis-open/blob/master/lwadtool/libadtool/ids.c#L183

File: nixos/modules/services/security/pbis.nix (hunk 2 above)
[the remark this answers is not in the document]
Comment (aanderse) [reply nixos/modules/services/security/pbis.nix]: :disappointed:

**Comment by aanderse** [comment]
@tekeri I was just intending to show you that having a stable `uid` (and `gid`) is possible with other solutions. As far as the SID... https://serverfault.com/questions/851864/get-sid-by-its-objectsid-using-ldapsearch

**Comment by tekeri** [comment]
@aanderse Thanks for the info. But I want to integrate NixOS into the existing network which is out of my control and using pbis there. uid inconsistency cause inconvenience with `rsync`, `ls` from container host, etc.
I've worked hard to avoid pbis and I'm still willing to migrate to any solution if it won't break.

**Comment by aanderse** [comment]
@tekeri yeah, makes sense to me. I was just trying to provide some information in case it provided any value. I think you've done a very good job on this module so far, which I imagine wasn't easy given the constraints of the software.

I'm happy to give this a test after someone else (possibly @Infinisil?) provides some more review on this.

:+1:

**Comment by tekeri** [comment]
@aanderse I'm happy to hear that, too.
Package is messy but I hope it should work like a charm :)

**Comment by tekeri** [comment]
ping, just for remind 😃

**Pull request closed without merging**